Add support for on-the-fly document upgrades

JKamsker opened the pull request:
This change would allow me to improve my library (https://github.com/JKamsker/LiteDb.Migration) significantly. The documents would then only be upgrades on load and not all at once.

pictos commented on LiteDB/Client/Database/LiteDatabase.cs:
```diff
@@ -32,19 +32,19 @@ public partial class LiteDatabase : ILiteDatabase
         /// </summary>
-        public LiteDatabase(string connectionString, BsonMapper mapper = null)
-            : this(new ConnectionString(connectionString), mapper)
+        public LiteDatabase(string connectionString, BsonMapper mapper = null, Action<EngineSettings> engineSettingsAction = null)
```
this is a breaking change, so in order to release it in the SR I suggest to add an overload to this ctor accepting this new Action

pictos commented on LiteDB/Engine/EngineSettings.cs:
```diff
@@ -66,6 +66,11 @@ public class EngineSettings
         public bool Upgrade { get; set; } = false;
 
+        /// <summary>
+        /// Is used to transform a bsonvalue from the database on read. This can be used to upgrade data from older versions.
```
```suggestion
         /// Is used to transform a <see cref="BsonValue"/> from the database on read. This can be used to upgrade data from older versions.
```

pictos commented on LiteDB/Engine/EngineState.cs:
```diff
@@ -51,5 +51,12 @@ public bool Handle(Exception ex)
+
+        public BsonValue ReadTransform(string collection, BsonValue value)
+        {
+            if (_settings?.ReadTransform == null) return value;
```
Using pattern-matching to check nullability is safer and better.

```suggestion
            if (_settings?.ReadTransform is null) return value;
```

pictos reviewed (commented):
Looking into the issue for more context, I think is better to @mbdavid to take a look on this and see if he wants to perform this change, or if using the `protected` methods are enough

JKamsker committed:
Add support for on-the-fly document upgrades

JKamsker committed:
Add deserialization hook to bsonmapper

JKamsker commented:
@pictos Thanks for the review, how did that slip through, that its a breaking change 😅 
Reverted the breaking changes and updated your requested changes.

So in my latest commit i added a deserialization hook.

Upsides:
- It is easier to add your own hook into that, because you only need a reference to the mapper
- Should be easier on the cpu because it only transforms those documents, which are actually needed.

Downsides:
- Queries may not run as expected, because there could be a transform between the query engine and the actual transform. But if that is a concern, the user should just use the ReadTransform in the engine settings.

JKamsker committed:
Fix indentation of csproj

pictos commented on LiteDB/LiteDB.csproj:
```diff
@@ -28,6 +28,7 @@
     <SignAssembly Condition="'$(OS)'=='Windows_NT'">true</SignAssembly>
     <AssemblyOriginatorKeyFile Condition="'$(Configuration)' == 'Release'">LiteDB.snk</AssemblyOriginatorKeyFile>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+    <LangVersion>9.0</LangVersion>
```
```suggestion
	<LangVersion>latest</LangVersion>
```

pictos reviewed (commented):
I'm away from my pc, so I can't test it right now

JKamsker committed:
Update LiteDB/LiteDB.csproj

Co-authored-by: Pedro Jesus <pedrojesus.cefet@gmail.com>

JKamsker committed:
Merge branch 'master' into ReadTransform

JKamsker commented:
Some tests failed before which are succeding (not the tests i touched)...sooo...

@pictos Can we merge? 😅
